### PR TITLE
Implement lock file; improve trap management

### PR DESCRIPTION
**Pull request by friendly-bits**
- Implement lock file check, creation, removal
- Create the trap only for actions which need it
- stop(): kill any running adblock-lean PID's
- status(): check the lock file and if it exists then report on it and exit
- status(): support additional return codes
- Only load config for actions which need it

### CODE REVIEW

File: adblock-lean [hunk 1]
```diff
@@ -915,10 +923,122 @@ update()
...
+		printf '%s\n' "${$} ${action}" > "${pid_file}" || { log_msg -err "Error: Failed to write to pid file '${pid_file}'."; exit 1; }
+	else
+		log_msg -err "Internal error: \${pid_file} variable is unset."
+		exit 1
+	fi
+	:
 }
 
-if [ "${action}" != "help" ] && [ "${action}" != "gen_config" ]
+rm_lock() {
```
Comment (lynxthecat) [inline adblock-lean:950]: Would you mind amending the functions to: 

x()
{
…
}

**Comment by lynxthecat** [comment]
Thanks for this. Looks pretty good from a very cursory read through whilst gardening.

File: adblock-lean (hunk 1 above)
Comment (friendly-bits) [reply adblock-lean]: > Would you mind amending the functions to

Done

**Comment by friendly-bits** [comment]
Changed the function declarations to
```
f ()
{
...
}
```

**Comment by friendly-bits** [comment]
Updated with slightly more compact variable assignment in `report_pid_action()`.

**Comment by friendly-bits** [comment]
^ Yet more tinkering with code compaction in `report_pid_action()` :)
I think that's it.

**Comment by friendly-bits** [comment]
^ Updated with more granular actions set in the lock file.
(and also slightly less spam in the log - we should consider reducing log spam further by utilizing the new `print_msg()` function for some of the less important messages, rather than `log_msg()`)

**Comment by friendly-bits** [comment]
^ Fixed incorrect variable name in `log_msg()` (was `_msg`, should be `msg`).

**Comment by lynxthecat** [comment]
Should be able to test soon. Maybe @Wizballs might also be able to test too.

**Comment by friendly-bits** [comment]
^
- mk_lock(): Only create the directory for the lock file if it doesn't exist yet (otherwise it could have errored out).
- mk_lock(): Remove trailing `.` from the action string when updating the action in the lock file.

**Comment by Wizballs** [comment]
Morning everyone, looks like there has been heaps going on!  Sure thing I'll give everything a test run.

**Comment by Wizballs** [comment]
@lynxthecat Tested multiple different times and service commands with the file lock.  I can't find any failure conditions, so I think this PR is good.

@friendly-bits That's really quite neat how it even outputs what action is currently being performed.  Also I saw another PR where stop/resume/pause was fixed.  Tested these also since I was here, all working well again.

**Comment by friendly-bits** [comment]
^
- Action stop: require cleanup (it wasn't removing `/var/run/adblock-lean/`)
- status(): swap return code 1 and return code 3 for consistency and so return code 1 means 'error' as per standard

(sorry, I found the issue with stop and then I also wanted to change the return codes of status. should be good to go now)

**Comment by friendly-bits** [comment]
^ Change function name from `advert_action` to `reg_action`.

**Comment by friendly-bits** [comment]
^
- Fix restart/reload when adblock-lean is running
- Fix the resume action when blocklist compression is enabled

**Comment by lynxthecat** [comment]
Seems to work well. This evidently required some careful thought and coding. Terrific job @friendly-bits!

**Pull request merged**